### DISCUSSION TITLE
Can RePack be used for setting up libraries?

### DISCUSSION THREAD

**Discussion by dmahajan980** [callstack/repack#311] [Q&A]
Hello, I would like to know if RePack can be used for bundling packages/libraries that can be consumed in other apps?

# Problem with [Builder Bob](https://github.com/callstack/react-native-builder-bob)

React Native Builder Bob (and Metro) does not support [subpath exports](https://nodejs.org/api/packages.html#subpath-exports). Metro has [released a new version](https://github.com/facebook/metro/releases/tag/v0.75.1) with experimental support for this feature but it isn't clear on when thus gets rolled out as stable and when Builder Bob accepts it.

Webpack, on the other hand, does support this and so I suppose RePack as well could since it's built on top of Webpack.

# Why it's needed?

In my package, I need this feature since we have multiple sub-modules in our package and would not like to expose them from the same entry point.

I am not sure if RePack can be used for bundling packages for consumption in other React Native apps. Any information on this would be highly appreciated. 

Thanks

**Comment by satya164** [comment]
I’m not well-versed in `exports` and ESM yet but from my understanding:

* It’s not accurate that Bob doesn’t support subpath exports, Bob just compiles the files. If the code is written to work with ESM (e.g. imports include extensions) then it should work after Bob compiles them as well. Though I would eventually like to add support for automatically compiling regular code to ESM compatible code.
* Webpack supporting `exports` is about consuming the lib that uses `exports` - I don’t think the library a mode Webpack takes regular code and produces ESM compatible code. At least no mention of `exports` here: https://webpack.js.org/guides/author-libraries/. You'd also not want to bundle the code to single bundle if you're trying to export multiple things under different paths.
* Webpack can be used to bundle libraries - as long as library imports marked as external and assets imports are left as `require`, it could be used to bundle RN libraries as well but I presume it’ll be very different from Repack
* `exports` isn’t the only way. Can also do this for subpaths https://github.com/callstack/react-native-paper/tree/main/react-navigation
